File: apps/android/app/src/main/java/com/verifd/android/data/BackendClient.kt
package com.verifd.android.data

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import com.verifd.android.network.RetryPolicy
import com.verifd.android.network.NetworkException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Backend API client with HMAC authentication for verifd.
 * Handles device registration, pass checking, and pass granting.
 */
class BackendClient private constructor(context: Context) {
    
    companion object {
        private const val TAG = "BackendClient"
        private const val PREFS_NAME = "verifd_device_auth"
        private const val KEY_DEVICE_ID = "device_id"
        private const val KEY_DEVICE_KEY = "device_key"
        private const val KEY_CUSTOM_URL = "custom_backend_url"
        
        @Volatile
        private var INSTANCE: BackendClient? = null
        
        fun getInstance(context: Context): BackendClient {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: BackendClient(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    private val prefs: SharedPreferences = try {
        // Use encrypted shared preferences for production
        val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
        EncryptedSharedPreferences.create(
            PREFS_NAME,
            masterKeyAlias,
            context,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    } catch (e: Exception) {
        Log.e(TAG, "Failed to create encrypted prefs, using regular", e)
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }
    
    private var deviceId: String? = prefs.getString(KEY_DEVICE_ID, null)
    private var deviceKey: String? = prefs.getString(KEY_DEVICE_KEY, null)
    
    // Use BuildConfig URL with optional runtime override
    private val baseUrl: String = prefs.getString(KEY_CUSTOM_URL, null) 
        ?: com.verifd.android.BuildConfig.BASE_URL
    
    // Retry policy for network requests
    private val retryPolicy = RetryPolicy(
        maxAttempts = 5,
        baseDelayMs = 500,
        maxDelayMs = 8000,
        jitterFactor = 0.2
    )
    
    /**
     * Set custom backend URL (for development/testing)
     */
    fun setCustomBackendUrl(url: String?) {
        if (url.isNullOrBlank()) {
            prefs.edit().remove(KEY_CUSTOM_URL).apply()
        } else {
            prefs.edit().putString(KEY_CUSTOM_URL, url).apply()
        }
        // Note: Requires app restart to take effect
    }
    
    /**
     * Get current backend URL (for debugging)
     */
    fun getCurrentBackendUrl(): String = baseUrl
    
    /**
     * Get or create device ID
     */
    private fun getOrCreateDeviceId(): String {
        return deviceId ?: run {
            val newId = generateDeviceId()
            deviceId = newId
            prefs.edit().putString(KEY_DEVICE_ID, newId).apply()
            newId
        }
    }
    
    /**
     * Generate a new device ID
     */
    private fun generateDeviceId(): String {
        val random = SecureRandom()
        val bytes = ByteArray(16)
        random.nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }
    
    /**
     * Register device with backend if not already registered
     */
    suspend fun ensureRegistered(): Boolean = withContext(Dispatchers.IO) {
        if (deviceId != null && deviceKey != null) {
            return@withContext true
        }
        
        try {
            val url = URL("$baseUrl/device/register")
            val connection = url.openConnection() as HttpURLConnection
            
            connection.apply {
                requestMethod = "POST"
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                connectTimeout = 10000
                readTimeout = 10000
            }
            
            val body = JSONObject().apply {
                put("platform", "android")
                put("model", android.os.Build.MODEL)
                put("app_version", "1.0.0") // TODO: Get from BuildConfig
            }
            
            connection.outputStream.use { os ->
                os.write(body.toString().toByteArray())
            }
            
            if (connection.responseCode == 200) {
                val response = connection.inputStream.bufferedReader().use { it.readText() }
                val json = JSONObject(response)
                
                if (json.getBoolean("success")) {
                    deviceId = json.getString("device_id")
                    deviceKey = json.getString("device_key")
                    
                    // Store credentials
                    prefs.edit().apply {
                        putString(KEY_DEVICE_ID, deviceId)
                        putString(KEY_DEVICE_KEY, deviceKey)
                        apply()
                    }
                    
                    Log.d(TAG, "Device registered: $deviceId")
                    return@withContext true
                }
            }
            
            Log.e(TAG, "Failed to register device: ${connection.responseCode}")
            false
            
        } catch (e: Exception) {
            Log.e(TAG, "Device registration error", e)
            false
        }
    }
    
    /**
     * Check if a phone number has an active pass
     */
    suspend fun checkPass(phoneNumber: String): PassCheckResult = withContext(Dispatchers.IO) {
        if (!ensureRegistered()) {
            return@withContext PassCheckResult.Error("Device not registered")
        }
        
        try {
            return@withContext retryPolicy.execute { attempt ->
                Log.d(TAG, "Checking pass, attempt $attempt")
                
                val encodedNumber = java.net.URLEncoder.encode(phoneNumber, "UTF-8")
                val url = URL("$baseUrl/pass/check?number_e164=$encodedNumber")
                val connection = url.openConnection() as HttpURLConnection
                
                connection.apply {
                    requestMethod = "GET"
                    connectTimeout = 5000
                    readTimeout = 5000
                }
                
                // Add HMAC authentication
                addHmacHeaders(connection, "GET", "/pass/check", "")
                
                val responseCode = connection.responseCode
                
                when {
                    responseCode == 200 -> {
                        val response = connection.inputStream.bufferedReader().use { it.readText() }
                        val json = JSONObject(response)
                        
                        if (json.getBoolean("allowed")) {
                            val scope = json.optString("scope", "24h")
                            val expiresAt = json.getString("expires_at")
                            val grantedTo = json.optString("granted_to_name", "Unknown")
                            
                            PassCheckResult.Allowed(
                                scope = scope,
                                expiresAt = expiresAt,
                                grantedToName = grantedTo
                            )
                        } else {
                            PassCheckResult.NotAllowed
                        }
                    }
                    responseCode == 429 -> {
                        // Rate limited - will be retried by RetryPolicy
                        throw NetworkException.fromHttpResponse(connection)
                    }
                    responseCode in 500..599 -> {
                        // Server error - will be retried
                        throw NetworkException.fromHttpResponse(connection)
                    }
                    else -> {
                        // Client error or not found
                        PassCheckResult.NotAllowed
                    }
                }
            }
        } catch (e: NetworkException) {
            if (e.statusCode == 429) {
                Log.w(TAG, "Rate limited after retries")
                PassCheckResult.RateLimited
            } else {
                Log.e(TAG, "Pass check failed after retries", e)
                PassCheckResult.Error(e.message ?: "Network error")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Pass check error", e)
            PassCheckResult.Error(e.message ?: "Network error")
        }
    }
    
    /**
     * Grant a pass to a phone number
     */
    suspend fun grantPass(
        phoneNumber: String,
        scope: String,
        name: String? = null,
        reason: String? = null
    ): GrantPassResult = withContext(Dispatchers.IO) {
        if (!ensureRegistered()) {
            return@withContext GrantPassResult.Error("Device not registered")
        }
        
        try {
            return@withContext retryPolicy.execute { attempt ->
                Log.d(TAG, "Granting pass, attempt $attempt")
                
                val url = URL("$baseUrl/passes/grant")
                val connection = url.openConnection() as HttpURLConnection
                
                connection.apply {
                    requestMethod = "POST"
                    doOutput = true
                    setRequestProperty("Content-Type", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }
                
                val body = JSONObject().apply {
                    put("number_e164", phoneNumber)
                    put("scope", scope)
                    name?.let { put("granted_to_name", it) }
                    reason?.let { put("reason", it) }
                    put("channel", "device")
                }
                
                val bodyString = body.toString()
                
                // Add HMAC authentication
                addHmacHeaders(connection, "POST", "/passes/grant", bodyString)
                
                connection.outputStream.use { os ->
                    os.write(bodyString.toByteArray())
                }
                
                val responseCode = connection.responseCode
                
                when {
                    responseCode == 200 -> {
                        val response = connection.inputStream.bufferedReader().use { it.readText() }
                        val json = JSONObject(response)
                        
                        if (json.getBoolean("success")) {
                            GrantPassResult.Success(
                                passId = json.getString("pass_id"),
                                expiresAt = json.getString("expires_at")
                            )
                        } else {
                            GrantPassResult.Error("Server returned success=false")
                        }
                    }
                    responseCode == 429 -> {
                        // Rate limited - will be retried by RetryPolicy
                        throw NetworkException.fromHttpResponse(connection)
                    }
                    responseCode in 500..599 -> {
                        // Server error - will be retried
                        throw NetworkException.fromHttpResponse(connection)
                    }
                    else -> {
                        // Client error - don't retry
                        GrantPassResult.Error("Failed to grant pass: $responseCode")
                    }
                }
            }
        } catch (e: NetworkException) {
            if (e.statusCode == 429) {
                Log.w(TAG, "Rate limited after retries")
                GrantPassResult.RateLimited
            } else {
                Log.e(TAG, "Grant pass failed after retries", e)
                GrantPassResult.Error(e.message ?: "Network error")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Grant pass error", e)
            GrantPassResult.Error(e.message ?: "Network error")
        }
    }
    
    /**
     * Add HMAC authentication headers to request
     */
    private fun addHmacHeaders(
        connection: HttpURLConnection,
        method: String,
        path: String,
        body: String
    ) {
        val timestamp = (System.currentTimeMillis() / 1000).toString()
        val payload = "$method:$path:$timestamp:$body"
        
        val signature = calculateHmac(payload, deviceKey!!)
        
        connection.setRequestProperty("X-Device-ID", deviceId)
        connection.setRequestProperty("X-Device-Sign", signature)
        connection.setRequestProperty("X-Timestamp", timestamp)
    }
    
    /**
     * Calculate HMAC-SHA256 signature
     */
    private fun calculateHmac(data: String, key: String): String {
        val secretKey = SecretKeySpec(key.toByteArray(), "HmacSHA256")
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(secretKey)
        val hmacBytes = mac.doFinal(data.toByteArray())
        return hmacBytes.joinToString("") { "%02x".format(it) }
    }
    
    // Result types
    sealed class PassCheckResult {
        data class Allowed(
            val scope: String,
            val expiresAt: String,
            val grantedToName: String
        ) : PassCheckResult()
        object NotAllowed : PassCheckResult()
        object RateLimited : PassCheckResult()
        data class Error(val message: String) : PassCheckResult()
    }
    
    /**
     * Fetch message templates for notification actions.
     * Uses caching on the backend (24h TTL per number).
     */
    suspend fun fetchMessageTemplates(
        phoneNumber: String,
        userName: String? = null,
        locale: String = "en-US"
    ): MessageTemplatesResult = withContext(Dispatchers.IO) {
        ensureRegistered()
        
        try {
            retryPolicy.execute { attempt ->
                val url = URL("${baseUrl}/v1/verify/link?" +
                    "phone_number=${phoneNumber.urlEncode()}" +
                    "&device_id=${getOrCreateDeviceId()}" +
                    "&locale=$locale" +
                    (userName?.let { "&user_name=${it.urlEncode()}" } ?: ""))
                
                val connection = url.openConnection() as HttpURLConnection
                connection.apply {
                    requestMethod = "GET"
                    connectTimeout = 5000
                    readTimeout = 5000
                }
                
                val responseCode = connection.responseCode
                
                when {
                    responseCode == 200 -> {
                        val response = connection.inputStream.bufferedReader().use { it.readText() }
                        val json = JSONObject(response)
                        
                        MessageTemplatesResult.Success(
                            smsTemplate = json.getString("sms_template"),
                            whatsAppTemplate = json.getString("whatsapp_template"),
                            verifyLink = json.getString("verify_link"),
                            cached = json.getBoolean("cached"),
                            ttlSeconds = json.getInt("ttl_seconds")
                        )
                    }
                    responseCode == 429 -> {
                        val response = connection.errorStream?.bufferedReader()?.use { it.readText() }
                        val json = response?.let { JSONObject(it) }
                        val retryAfter = json?.optInt("retry_after", 60) ?: 60
                        
                        MessageTemplatesResult.RateLimited(retryAfter)
                    }
                    responseCode == 400 -> {
                        MessageTemplatesResult.Error("Invalid phone number")
                    }
                    responseCode in 500..599 -> {
                        // Server error - will be retried
                        throw NetworkException.fromHttpResponse(connection)
                    }
                    else -> {
                        MessageTemplatesResult.Error("Failed to fetch templates: $responseCode")
                    }
                }
            }
        } catch (e: NetworkException) {
            if (e.statusCode == 429) {
                Log.w(TAG, "Rate limited fetching templates")
                MessageTemplatesResult.RateLimited(60)
            } else {
                Log.e(TAG, "Failed to fetch templates", e)
                MessageTemplatesResult.Error(e.message ?: "Network error")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to fetch templates", e)
            MessageTemplatesResult.Error(e.message ?: "Unknown error")
        }
    }
    
    private fun String.urlEncode(): String {
        return java.net.URLEncoder.encode(this, "UTF-8")
    }
    
    sealed class MessageTemplatesResult {
        data class Success(
            val smsTemplate: String,
            val whatsAppTemplate: String,
            val verifyLink: String,
            val cached: Boolean,
            val ttlSeconds: Int
        ) : MessageTemplatesResult()
        data class RateLimited(val retryAfterSeconds: Int) : MessageTemplatesResult()
        data class Error(val message: String) : MessageTemplatesResult()
    }
    
    sealed class GrantPassResult {
        data class Success(
            val passId: String,
            val expiresAt: String
        ) : GrantPassResult()
        object RateLimited : GrantPassResult()
        data class Error(val message: String) : GrantPassResult()
    }
}

File: apps/ios/verifd/Services/BackendClient.swift
import Foundation
import CryptoKit
import Security

/**
 * Backend API client with HMAC authentication for verifd iOS.
 * Handles device registration, pass checking, and pass granting.
 */
class BackendClient {
    
    // MARK: - Singleton
    static let shared = BackendClient()
    
    // MARK: - Constants
    private let keychainService = "com.verifd.ios"
    private let deviceIdKey = "device_id"
    private let deviceKeyKey = "device_key"
    private let customUrlKey = "custom_backend_url"
    
    // MARK: - Retry Policy
    private let retryPolicy = RetryPolicy(
        maxAttempts: 5,
        baseDelayMs: 500,
        maxDelayMs: 8000,
        jitterFactor: 0.2
    )
    
    // MARK: - Base URL Configuration
    private var baseURL: String {
        // Check for runtime override first
        if let customUrl = UserDefaults.standard.string(forKey: customUrlKey),
           !customUrl.isEmpty {
            return customUrl
        }
        
        // Fall back to build configuration
        if let bundleUrl = Bundle.main.object(forInfoDictionaryKey: "BackendURL") as? String,
           !bundleUrl.isEmpty {
            return bundleUrl
        }
        
        // Default fallback for development
        return "http://localhost:3000"
    }
    
    // MARK: - Properties
    private var deviceId: String? {
        get { loadFromKeychain(key: deviceIdKey) }
        set { saveToKeychain(key: deviceIdKey, value: newValue) }
    }
    
    private var deviceKey: String? {
        get { loadFromKeychain(key: deviceKeyKey) }
        set { saveToKeychain(key: deviceKeyKey, value: newValue) }
    }
    
    private init() {}
    
    // MARK: - Configuration
    
    func setCustomBackendURL(_ url: String?) {
        if let url = url, !url.isEmpty {
            UserDefaults.standard.set(url, forKey: customUrlKey)
        } else {
            UserDefaults.standard.removeObject(forKey: customUrlKey)
        }
        // Note: Requires app restart to take effect
    }
    
    func getCurrentBackendURL() -> String {
        return baseURL
    }
    
    func getBuildVariant() -> String {
        return Bundle.main.object(forInfoDictionaryKey: "BuildVariant") as? String ?? "unknown"
    }
    
    // MARK: - Device Registration
    
    func ensureRegistered() async throws -> Bool {
        if deviceId != nil && deviceKey != nil {
            return true
        }
        
        let url = URL(string: "\(baseURL)/device/register")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10.0
        
        let body: [String: Any] = [
            "platform": "ios",
            "model": UIDevice.current.model,
            "app_version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            print("Failed to register device: \((response as? HTTPURLResponse)?.statusCode ?? -1)")
            return false
        }
        
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        
        if let success = json?["success"] as? Bool, success,
           let deviceId = json?["device_id"] as? String,
           let deviceKey = json?["device_key"] as? String {
            self.deviceId = deviceId
            self.deviceKey = deviceKey
            print("Device registered: \(deviceId)")
            return true
        }
        
        return false
    }
    
    // MARK: - Pass Checking
    
    enum PassCheckResult {
        case allowed(scope: String, expiresAt: String, grantedToName: String)
        case notAllowed
        case rateLimited
        case error(String)
    }
    
    func checkPass(phoneNumber: String) async -> PassCheckResult {
        guard try? await ensureRegistered() else {
            return .error("Device not registered")
        }
        
        do {
            return try await retryPolicy.execute { attempt in
                print("Checking pass, attempt \(attempt)")
                
                let encodedNumber = phoneNumber.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? phoneNumber
                let url = URL(string: "\(baseURL)/pass/check?number_e164=\(encodedNumber)")!
                var request = URLRequest(url: url)
                request.httpMethod = "GET"
                request.timeoutInterval = 5.0
                
                // Add HMAC authentication
                addHmacHeaders(to: &request, method: "GET", path: "/pass/check", body: "")
                
                let (data, response) = try await URLSession.shared.data(for: request)
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw NetworkError.invalidResponse
                }
                
                switch httpResponse.statusCode {
                case 200:
                    let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
                    
                    if let allowed = json?["allowed"] as? Bool, allowed {
                        let scope = json?["scope"] as? String ?? "24h"
                        let expiresAt = json?["expires_at"] as? String ?? ""
                        let grantedTo = json?["granted_to_name"] as? String ?? "Unknown"
                        
                        return PassCheckResult.allowed(scope: scope, expiresAt: expiresAt, grantedToName: grantedTo)
                    } else {
                        return PassCheckResult.notAllowed
                    }
                case 429:
                    // Rate limited - will be retried by RetryPolicy
                    throw NetworkError.fromHTTPResponse(httpResponse)
                case 500...599:
                    // Server error - will be retried
                    throw NetworkError.fromHTTPResponse(httpResponse)
                default:
                    // Client error or not found
                    return PassCheckResult.notAllowed
                }
            }
        } catch {
            if case NetworkError.httpError(429, _) = error {
                print("Rate limited after retries")
                return .rateLimited
            } else {
                print("Pass check failed after retries: \(error)")
                return .error(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Pass Granting
    
    enum GrantPassResult {
        case success(passId: String, expiresAt: String)
        case rateLimited
        case error(String)
    }
    
    func grantPass(phoneNumber: String, scope: String, name: String? = nil, reason: String? = nil) async -> GrantPassResult {
        guard try? await ensureRegistered() else {
            return .error("Device not registered")
        }
        
        do {
            return try await retryPolicy.execute { attempt in
                print("Granting pass, attempt \(attempt)")
                
                let url = URL(string: "\(baseURL)/passes/grant")!
                var request = URLRequest(url: url)
                request.httpMethod = "POST"
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.timeoutInterval = 10.0
                
                var body: [String: Any] = [
                    "number_e164": phoneNumber,
                    "scope": scope,
                    "channel": "device"
                ]
                
                if let name = name {
                    body["granted_to_name"] = name
                }
                
                if let reason = reason {
                    body["reason"] = reason
                }
                
                let bodyData = try JSONSerialization.data(withJSONObject: body)
                let bodyString = String(data: bodyData, encoding: .utf8) ?? "{}"
                
                // Add HMAC authentication
                addHmacHeaders(to: &request, method: "POST", path: "/passes/grant", body: bodyString)
                
                request.httpBody = bodyData
                
                let (data, response) = try await URLSession.shared.data(for: request)
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw NetworkError.invalidResponse
                }
                
                switch httpResponse.statusCode {
                case 200:
                    let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
                    
                    if let success = json?["success"] as? Bool, success,
                       let passId = json?["pass_id"] as? String,
                       let expiresAt = json?["expires_at"] as? String {
                        return GrantPassResult.success(passId: passId, expiresAt: expiresAt)
                    } else {
                        return GrantPassResult.error("Server returned success=false")
                    }
                case 429:
                    // Rate limited - will be retried by RetryPolicy
                    throw NetworkError.fromHTTPResponse(httpResponse)
                case 500...599:
                    // Server error - will be retried
                    throw NetworkError.fromHTTPResponse(httpResponse)
                default:
                    // Client error - don't retry
                    return GrantPassResult.error("Failed to grant pass: \(httpResponse.statusCode)")
                }
            }
        } catch {
            if case NetworkError.httpError(429, _) = error {
                print("Rate limited after retries")
                return .rateLimited
            } else {
                print("Grant pass failed after retries: \(error)")
                return .error(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Pass Sync
    
    struct Pass {
        let id: String
        let numberE164: String
        let grantedBy: String
        let grantedToName: String
        let reason: String?
        let expiresAt: String
        let createdAt: String
    }
    
    func syncPassesSince(timestamp: TimeInterval) async -> Result<[Pass], Error> {
        guard try? await ensureRegistered() else {
            return .failure(NSError(domain: "BackendClient", code: -1, userInfo: [NSLocalizedDescriptionKey: "Device not registered"]))
        }
        
        do {
            let url = URL(string: "\(baseURL)/passes/since?ts=\(Int(timestamp))")!
            var request = URLRequest(url: url)
            request.httpMethod = "GET"
            request.timeoutInterval = 10.0
            
            // Add HMAC authentication
            addHmacHeaders(to: &request, method: "GET", path: "/passes/since", body: "")
            
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return .failure(NSError(domain: "BackendClient", code: -1, userInfo: [NSLocalizedDescriptionKey: "HTTP error: \((response as? HTTPURLResponse)?.statusCode ?? -1)"]))
            }
            
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            
            if let success = json?["success"] as? Bool, success,
               let passesData = json?["passes"] as? [[String: Any]] {
                
                let passes = passesData.compactMap { passDict -> Pass? in
                    guard let id = passDict["id"] as? String,
                          let numberE164 = passDict["number_e164"] as? String,
                          let grantedBy = passDict["granted_by"] as? String,
                          let grantedToName = passDict["granted_to_name"] as? String,
                          let expiresAt = passDict["expires_at"] as? String,
                          let createdAt = passDict["created_at"] as? String else {
                        return nil
                    }
                    
                    return Pass(
                        id: id,
                        numberE164: numberE164,
                        grantedBy: grantedBy,
                        grantedToName: grantedToName,
                        reason: passDict["reason"] as? String,
                        expiresAt: expiresAt,
                        createdAt: createdAt
                    )
                }
                
                return .success(passes)
            }
            
            return .failure(NSError(domain: "BackendClient", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response format"]))
            
        } catch {
            return .failure(error)
        }
    }
    
    // MARK: - HMAC Authentication
    
    private func addHmacHeaders(to request: inout URLRequest, method: String, path: String, body: String) {
        guard let deviceId = deviceId, let deviceKey = deviceKey else { return }
        
        let timestamp = String(Int(Date().timeIntervalSince1970))
        let payload = "\(method):\(path):\(timestamp):\(body)"
        
        let signature = calculateHmac(data: payload, key: deviceKey)
        
        request.setValue(deviceId, forHTTPHeaderField: "X-Device-ID")
        request.setValue(signature, forHTTPHeaderField: "X-Device-Sign")
        request.setValue(timestamp, forHTTPHeaderField: "X-Timestamp")
    }
    
    private func calculateHmac(data: String, key: String) -> String {
        let keyData = Data(key.utf8)
        let dataToSign = Data(data.utf8)
        
        let signature = HMAC<SHA256>.authenticationCode(for: dataToSign, using: SymmetricKey(data: keyData))
        return Data(signature).map { String(format: "%02x", $0) }.joined()
    }
    
    // MARK: - Keychain Storage
    
    private func saveToKeychain(key: String, value: String?) {
        guard let value = value else {
            // Delete if nil
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrService as String: keychainService,
                kSecAttrAccount as String: key
            ]
            SecItemDelete(query as CFDictionary)
            return
        }
        
        let data = value.data(using: .utf8)!
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        
        // Try to update first
        let updateQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: key
        ]
        
        let status = SecItemUpdate(updateQuery as CFDictionary, [kSecValueData as String: data] as CFDictionary)
        
        if status == errSecItemNotFound {
            // Add new item
            SecItemAdd(query as CFDictionary, nil)
        }
    }
    
    private func loadFromKeychain(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess,
           let data = result as? Data,
           let value = String(data: data, encoding: .utf8) {
            return value
        }
        
        return nil
    }
}